File: backend/app/schemas/rbac.py
"""Request/response schemas for User & Role administration (RBAC - Phase 3).

Admin responses carry only **safe administrative metadata** - never a password
hash, token or any auth secret (those fields do not exist on any model here).
Permission *codes* are stable machine identifiers and are never translated; the
frontend renders a Spanish label next to the code.
"""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.rbac import DESCRIPTION_MAX_LENGTH, ROLE_NAME_MAX_LENGTH

# Page-size conventions mirror the rest of the API.
USERS_DEFAULT_PAGE_SIZE = 20
ROLES_DEFAULT_PAGE_SIZE = 50
MAX_PAGE_SIZE = 100


# --------------------------------------------------------------------------
# Shared refs
# --------------------------------------------------------------------------


class RoleRef(BaseModel):
    id: uuid.UUID
    name: str
    slug: str
    is_system: bool

    model_config = ConfigDict(from_attributes=True)


class PermissionRead(BaseModel):
    code: str
    group: str
    description: str


class PermissionCatalog(BaseModel):
    """The full permission catalog, grouped for the matrix editor."""

    groups: list[str]
    permissions: list[PermissionRead]


# --------------------------------------------------------------------------
# Users
# --------------------------------------------------------------------------


class AdminUserListItem(BaseModel):
    id: uuid.UUID
    email: str
    #: Lifecycle state - ``pending`` / ``active`` / ``rejected`` / ``disabled``.
    account_status: str
    #: Derived (``account_status == active``); kept for convenience.
    is_active: bool
    created_at: datetime
    roles: list[RoleRef]


class AdminUserPage(BaseModel):
    items: list[AdminUserListItem]
    page: int = Field(ge=1)
    page_size: int = Field(ge=1, le=MAX_PAGE_SIZE)
    total: int = Field(ge=0)
    total_pages: int = Field(ge=0)


class AdminUserDetail(BaseModel):
    id: uuid.UUID
    email: str
    account_status: str
    is_active: bool
    created_at: datetime
    updated_at: datetime
    roles: list[RoleRef]
    #: Union across assigned roles. For a non-active account these permissions
    #: are **not** in effect - the UI shows them as "will apply once approved /
    #: reactivated" and the backend rejects the account regardless.
    permissions: list[str]
    #: True when this user is the only active Administrator (drives self-lockout
    #: guidance in the UI).
    is_last_active_admin: bool


class AdminUserUpdate(BaseModel):
    """Runtime enable / disable of an already-provisioned account
    (``active`` <-> ``disabled``). Pending / rejected requests go through the
    approve / reject endpoints instead."""

    model_config = ConfigDict(extra="forbid")

    is_active: bool


class UserRolesUpdate(BaseModel):
    """Replace a user's role set with exactly ``role_ids`` (de-duplicated)."""

    model_config = ConfigDict(extra="forbid")

    role_ids: list[uuid.UUID] = Field(default_factory=list, max_length=50)


class ApproveAccessRequest(BaseModel):
    """Approve a pending (or previously rejected) access request. **At least one
    role is required** - approving with none creates an unusable account."""

    model_config = ConfigDict(extra="forbid")

    role_ids: list[uuid.UUID] = Field(min_length=1, max_length=50)


# --------------------------------------------------------------------------
# Roles
# --------------------------------------------------------------------------


class RoleListItem(BaseModel):
    id: uuid.UUID
    name: str
    slug: str
    description: str | None
    is_system: bool
    user_count: int
    permission_count: int


class RolePage(BaseModel):
    items: list[RoleListItem]
    total: int = Field(ge=0)


class RoleUserRef(BaseModel):
    id: uuid.UUID
    email: str
    is_active: bool

    model_config = ConfigDict(from_attributes=True)


class RoleDetail(BaseModel):
    id: uuid.UUID
    name: str
    slug: str
    description: str | None
    is_system: bool
    created_at: datetime
    updated_at: datetime
    permissions: list[str]
    users: list[RoleUserRef]


class RoleCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=ROLE_NAME_MAX_LENGTH)
    description: str | None = Field(default=None, max_length=DESCRIPTION_MAX_LENGTH)
    permissions: list[str] = Field(default_factory=list, max_length=100)


class RoleUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=ROLE_NAME_MAX_LENGTH)
    description: str | None = Field(default=None, max_length=DESCRIPTION_MAX_LENGTH)


class RolePermissionsUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    permissions: list[str] = Field(default_factory=list, max_length=100)

File: frontend/src/types/rbac.ts
/**
 * User & Role administration types (RBAC - Phase 3).
 *
 * Permission *codes* are stable machine identifiers, never translated. The
 * network response is untrusted - every shape below has a runtime guard.
 */

import { PERMISSION_GROUPS, PERMISSIONS } from "@/lib/permissions";
import type { AccountStatus } from "@/types/auth";

export type { AccountStatus };

export interface RoleRef {
  id: string;
  name: string;
  slug: string;
  is_system: boolean;
}

export interface AdminUserListItem {
  id: string;
  email: string;
  account_status: AccountStatus;
  is_active: boolean;
  created_at: string;
  roles: RoleRef[];
}

export interface AdminUserPage {
  items: AdminUserListItem[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

export interface AdminUserDetail {
  id: string;
  email: string;
  account_status: AccountStatus;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  roles: RoleRef[];
  permissions: string[];
  is_last_active_admin: boolean;
}

export interface RoleListItem {
  id: string;
  name: string;
  slug: string;
  description: string | null;
  is_system: boolean;
  user_count: number;
  permission_count: number;
}

export interface RolePage {
  items: RoleListItem[];
  total: number;
}

export interface RoleUserRef {
  id: string;
  email: string;
  is_active: boolean;
}

export interface RoleDetail {
  id: string;
  name: string;
  slug: string;
  description: string | null;
  is_system: boolean;
  created_at: string;
  updated_at: string;
  permissions: string[];
  users: RoleUserRef[];
}

export interface PermissionRead {
  code: string;
  group: string;
  description: string;
}

export interface PermissionCatalog {
  groups: string[];
  permissions: PermissionRead[];
}

// --- runtime guards -------------------------------------------------------

function isRecord(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function isRoleRef(v: unknown): v is RoleRef {
  if (!isRecord(v)) return false;
  return (
    typeof v.id === "string" &&
    typeof v.name === "string" &&
    typeof v.slug === "string" &&
    typeof v.is_system === "boolean"
  );
}

function isStringArray(v: unknown): v is string[] {
  return Array.isArray(v) && v.every((x) => typeof x === "string");
}

const ACCOUNT_STATUSES: readonly string[] = ["pending", "active", "rejected", "disabled"];
function isAccountStatus(v: unknown): v is AccountStatus {
  return typeof v === "string" && ACCOUNT_STATUSES.includes(v);
}

export function isAdminUserListItem(v: unknown): v is AdminUserListItem {
  if (!isRecord(v)) return false;
  return (
    typeof v.id === "string" &&
    typeof v.email === "string" &&
    isAccountStatus(v.account_status) &&
    typeof v.is_active === "boolean" &&
    typeof v.created_at === "string" &&
    Array.isArray(v.roles) &&
    v.roles.every(isRoleRef)
  );
}

export function isAdminUserPage(v: unknown): v is AdminUserPage {
  if (!isRecord(v)) return false;
  return (
    Array.isArray(v.items) &&
    v.items.every(isAdminUserListItem) &&
    typeof v.page === "number" &&
    typeof v.total === "number" &&
    typeof v.total_pages === "number"
  );
}

export function isAdminUserDetail(v: unknown): v is AdminUserDetail {
  if (!isRecord(v)) return false;
  return (
    typeof v.id === "string" &&
    typeof v.email === "string" &&
    isAccountStatus(v.account_status) &&
    typeof v.is_active === "boolean" &&
    Array.isArray(v.roles) &&
    v.roles.every(isRoleRef) &&
    isStringArray(v.permissions) &&
    typeof v.is_last_active_admin === "boolean"
  );
}

export function isRoleListItem(v: unknown): v is RoleListItem {
  if (!isRecord(v)) return false;
  return (
    typeof v.id === "string" &&
    typeof v.name === "string" &&
    typeof v.slug === "string" &&
    typeof v.is_system === "boolean" &&
    typeof v.user_count === "number" &&
    typeof v.permission_count === "number"
  );
}

export function isRolePage(v: unknown): v is RolePage {
  if (!isRecord(v)) return false;
  return Array.isArray(v.items) && v.items.every(isRoleListItem) && typeof v.total === "number";
}

export function isRoleDetail(v: unknown): v is RoleDetail {
  if (!isRecord(v)) return false;
  return (
    typeof v.id === "string" &&
    typeof v.name === "string" &&
    typeof v.slug === "string" &&
    typeof v.is_system === "boolean" &&
    isStringArray(v.permissions) &&
    Array.isArray(v.users)
  );
}

export function isPermissionCatalog(v: unknown): v is PermissionCatalog {
  if (!isRecord(v)) return false;
  return (
    isStringArray(v.groups) &&
    Array.isArray(v.permissions) &&
    v.permissions.every(
      (p) =>
        isRecord(p) &&
        typeof p.code === "string" &&
        typeof p.group === "string" &&
        typeof p.description === "string",
    )
  );
}

export function isRoleRefArray(v: unknown): v is RoleRef[] {
  return Array.isArray(v) && v.every(isRoleRef);
}

/** Split a code list into `{ group: codes[] }`, ordered by the catalog. */
export function groupPermissions(codes: readonly string[]): Record<string, string[]> {
  const out: Record<string, string[]> = {};
  for (const group of PERMISSION_GROUPS) {
    const inGroup = PERMISSIONS.filter(
      (p) => p.startsWith(`${group}.`) && codes.includes(p),
    );
    if (inGroup.length) out[group] = inGroup;
  }
  return out;
}
